frame/staking/src/tests.rs

```diff
@@ -2702,29 +2693,1028 @@ fn remove_multi_deferred() {
+		voter = 5;
+		bond_nominator(voter, 1000 + voter, 100, vec![20, 30, 40]);
+	}
+
+	fn offchainify(ext: &mut TestExternalities) -> Arc<RwLock<PoolState>> {
+		let (offchain, _state) = TestOffchainExt::new();
+		let (pool, state) = TestTransactionPoolExt::new();
+		let keystore = KeyStore::new();
+
+		let account = LOCAL_KEY_ACCOUNT.with(|v| *v.borrow());
+		let key = dummy_sr25519::dummy_key_for(account);
+		let _ = keystore
+			.write()
+			.insert_unknown(
+				<DummyT as sp_application_crypto::AppKey>::ID,
+				&format!("{}/staking{}", mock::PHRASE, account),
+				key.as_ref(),
+			)
+			.unwrap();
+		debug::native::debug!(
+			target: "staking",
+			"generated key for account {}: {:?}",
+			account,
+			key,
+		);
+		ext.register_extension(OffchainExt::new(offchain));
+		ext.register_extension(TransactionPoolExt::new(pool));
+		ext.register_extension(KeystoreExt(keystore));
+		state
+	}
+
+	#[test]
+	fn is_current_session_final_works() {
+		ExtBuilder::default()
+			.session_per_era(3)
+			.build()
+			.execute_with(|| {
+				start_era(1);
+				assert_eq!(Session::current_index(), 3);
+				assert_eq!(Staking::current_era(), Some(1));
+				assert_eq!(Staking::is_current_session_final(), false);
+
+				start_session(4);
+				assert_eq!(Session::current_index(), 4);
+				assert_eq!(Staking::current_era(), Some(1));
+				assert_eq!(Staking::is_current_session_final(), true);
+
+				start_session(5);
+				assert_eq!(Session::current_index(), 5);
+				// era changed.
+				assert_eq!(Staking::current_era(), Some(2));
+				assert_eq!(Staking::is_current_session_final(), false);
+			})
+	}
+
+	#[test]
+	fn offchain_election_flag_is_triggered() {
+		ExtBuilder::default()
+			.session_per_era(5)
+			.session_length(10)
+			.election_lookahead(3)
+			.build()
+			.execute_with(|| {
+				run_to_block(7);
+				assert_session_era!(0, 0);
+
+				run_to_block(10);
+				assert_session_era!(1, 0);
+				assert_eq!(Staking::era_election_status(), ElectionStatus::Closed);
+				assert!(Staking::snapshot_nominators().is_none());
+				assert!(Staking::snapshot_validators().is_none());
+
+				run_to_block(18);
+				assert_session_era!(1, 0);
+
+				run_to_block(27);
+				assert_session_era!(2, 0);
+
+				run_to_block(36);
+				assert_session_era!(3, 0);
+
+				// fist era has session 0, which has 0 blocks length, so we have in total 40 blocks
+				// in the era.
+				run_to_block(37);
+				assert_session_era!(3, 0);
+				assert_eq!(Staking::era_election_status(), ElectionStatus::Open(37));
+				assert!(Staking::snapshot_nominators().is_some());
+				assert!(Staking::snapshot_validators().is_some());
+
+				run_to_block(38);
+				assert_eq!(Staking::era_election_status(), ElectionStatus::Open(37));
+
+				run_to_block(39);
+				assert_eq!(Staking::era_election_status(), ElectionStatus::Open(37));
+
+				run_to_block(40);
+				assert_session_era!(4, 0); // TODO @guillaume I was expecting active era to be bumped by this point. Double check
```

Comment:
active_era will start at block 50.
block 1: new_session(0); new_session(1); start_session(0); active_era==0; current_era==0;
block 10: new_session(2); start_session(1);
...
block 40: new_session(5); start_session(4); active_era==0; current_era==1;
(the staking::current_era is delayed by one session by session module.)
block 50: new_session(6); start_session(5); active_era==1; current_era==1;

```suggestion
				assert_session_era!(4, 0);
```